### ISSUE TITLE
[BUG] Assigning `struct` with `tuple`

### ISSUE THREAD

**Issue by jakirkham** [cython/cython#3806]
**Describe the bug**

Currently it does not appear to be possible to assign a `tuple` (or other sequence) to a Cython `struct`. It would be useful to perform this kind of assignment for succinctness.  It also aligns with the behavior of [`ctuple`s]( https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html?highlight=ctuple#types ), which also use `struct`s under-the-hood.

**To Reproduce**

Code to reproduce the behavior:

```cython
cdef struct bar_t:
    double d
    int i

cdef bar_t bar = (3.0, 2)  # <--- compiler error here
```

1st Workaround (use `dict`):

```cython
cdef struct bar_t:
    double d
    int i

cdef bar_t bar = {"d": 3.0, "i": 2}
```

2nd Workaround (use `ctuple`):

```cython
ctypedef (double, int) bar_t

cdef bar_t bar = (3.0, 2)
```

**Expected behavior**

It would be useful to allow sequences (like `tuple`s) to be coerced to `struct`s analogous to how `ctuple`s allow this behavior.

**Environment (please complete the following information):**
 - OS: macOS
 - Python version 3.8.5
 - Cython version 0.29.21

**Additional context**

NA

**Comment by scoder** [comment]
You can already write
```cython
b = bar_t(3.0, 2)
```

This request seems reasonable for non-literals, though. C structs have ordered attributes, we can allow an assignment from an ordered iterable, similar to how we support assignments from dicts. But it would also mean that we'd have to generate a lot more code for a simple statement like
```cython
cdef bar_t b = x
```
since we can't know whether `x` will be an iterable, sequence or dict at runtime. So we'd end up having to support all three in the C code (or, at least two, but sequences are faster than iterables, so we'd still want all three). Difficult to say if that is worth it.

**Comment by jakirkham** [comment]
Yeah in the use case where this came up, we know there is a `tuple`, but we don't know its contents. So I suppose we could do things like this...

```cython
cdef tuple o = ...
cdef bar_t v = bar_t(o[0], o[1])
```

Wound up going with something like this for now...

```cython
cdef tuple o = ...
cdef bar_t v
v.d, v.i = o
```

**Comment by TeamSpen210** [comment]
What about potentially supporting `bar_t(*iterable)` or `bar_t(**dict)`? Then you can explicitly request either conversion type.

**Comment by jakirkham** [comment]
Yeah those would be good as well. The first doesn’t work today though the latter works just through assignment as noted in the OP.

**Comment by scoder** [comment]
> What about potentially supporting `bar_t(*iterable)` or `bar_t(**dict)`? Then you can explicitly request either conversion type.

I think this is a good way to do it, yes. Can probably be done by (internally) spelling out the first argument list as index access with the fixed number of struct fields, and the second as a cast to the struct type (which already works).

One issue: for nested structs, sequences that contain themselves sequences would then still fail to assign since there is no way to make their way of conversion explicit. It's equally likely to find a dict in a tuple item as another tuple.
